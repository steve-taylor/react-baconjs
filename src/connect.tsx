import React from 'react'

import useWidgetState from './useWidgetState'
import {WidgetContext} from './types'

/**
 * HOC to render a component and optional loading state component with widget state injected.
 *
 * This provides a simple way to decouple React components from widget state. Unlike Reduxâ€™s
 * connect, the only supported transformation of state to props is to optionally select a
 * subset of state object properties.
 *
 * Status: Experimental and subject to API change
 *
 * @param context - widget context
 * @param keys    - optional subset of state to use
 */
export default function connect<STATE>(context: React.Context<WidgetContext<STATE>>, keys?: (keyof STATE)[]) {
    // eslint-disable-next-line react/display-name
    return (component: React.ComponentType, loadingComponent: React.ComponentType): JSX.Element | null => (
        <Connector {...{context, compare: keys, component, loadingComponent}} />
    )
}

type Props<STATE> = {
    context: React.Context<WidgetContext<STATE>>
    keys?: (keyof STATE)[]
    component: React.ComponentType
    loadingComponent?: React.ComponentType
}

function Connector<STATE>({
    context,
    keys,
    component: Component,
    loadingComponent: LoadingComponent,
}: Props<STATE>): JSX.Element | null {
    const [state, isLoading] = useWidgetState<STATE>(context, keys)

    return !isLoading ? (
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        <Component {...(keys ? keys.reduce((acc, key) => Object.assign(acc, {[key]: state![key]}), {}) : state!)} />
    ) : LoadingComponent ? (
        <LoadingComponent />
    ) : null
}
