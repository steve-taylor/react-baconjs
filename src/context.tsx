import {Property} from 'baconjs';
import React from 'react';

import {BaseProps, Serializable, SerializableObject, WidgetData} from './types';

export interface ServerContextType<S, H extends Serializable, M> {
    getStream: (key: string) => Property<WidgetData<S, H, M>>;
    registerStream?: (key: string, stream$: Property<WidgetData<S, H, M>>) => void;
    onMeta?: (meta: M) => void;
    onError?: (error: unknown) => void;
}

const defaultServerContext: ServerContextType<unknown, Serializable, unknown> = {
    getStream: () => {
        throw new Error('<ServerContext.Provider /> is missing from the current scope.');
    },
};

export type HydrationContextType<
    P extends BaseProps,
    H extends Serializable,
> = (name: string, props: P) => {hydration: H | undefined; elementId: string | undefined;};

const defaultHydrationContext: HydrationContextType<SerializableObject, Serializable> = () => {
    throw new Error('<HydrationContext.Provider /> is missing from the current scope.');
}

export const HYDRATION: unique symbol = Symbol('hydration');
export const SERVER = Symbol('ssr');

export type Phase = typeof HYDRATION | typeof SERVER | null;
export type PhaseContextType = () => Phase;

export const ServerContext = React.createContext<ServerContextType<unknown, Serializable, unknown>>(defaultServerContext);

export const HydrationContext = React.createContext<HydrationContextType<SerializableObject, Serializable>>(defaultHydrationContext);

export const PhaseContext = React.createContext<PhaseContextType>(() => null);

ServerContext.displayName = 'ServerContext';
HydrationContext.displayName = 'HydrationContext';
PhaseContext.displayName = 'PhaseContext';

export function ServerContextConsumer<S, H extends Serializable, M>({
    children,
}: {
    children: (value: ServerContextType<S, H, M>) => React.ReactNode,
}): JSX.Element {
    const Ctx = ServerContext as unknown as React.Context<ServerContextType<S, H, M>>;

    return (
        <Ctx.Consumer>
            {children}
        </Ctx.Consumer>
    );
}

export function ServerContextProvider<S, H extends Serializable, M>({
    value,
    children,
}: {
    value: ServerContextType<S, H, M>,
    children: React.ReactNode,
}): JSX.Element {
    const Ctx = ServerContext as unknown as React.Context<ServerContextType<S, H, M>>;

    return (
        <Ctx.Provider value={value}>
            {children}
        </Ctx.Provider>
    );
}

export function HydrationContextConsumer<P extends BaseProps, H extends Serializable>({
    children,
}: {
    children: (value: HydrationContextType<P, H>) => React.ReactNode,
}): JSX.Element {
    const Ctx = HydrationContext as unknown as React.Context<HydrationContextType<P, H>>;

    return (
        <Ctx.Consumer>
            {children}
        </Ctx.Consumer>
    );
}

export function HydrationContextProvider<P extends BaseProps, H extends Serializable>({
    value,
    children,
}: {
    value: HydrationContextType<P, H>,
    children: React.ReactNode,
}): JSX.Element {
    const Ctx = HydrationContext as unknown as React.Context<HydrationContextType<P, H>>;

    return (
        <Ctx.Provider value={value}>
            {children}
        </Ctx.Provider>
    );
}
