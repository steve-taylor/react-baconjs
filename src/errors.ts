export const SSR_TIMEOUT_ERROR = Symbol('timeout')

export function noImmediateStateOnServerError(name: string): string {
    return `An unexpected error occurred while server-side rendering “${name || '(unknown)'}”`
}

export function noContext(): string {
    return 'Cannot use Inject or useWidgetState outside the scope of the specified context’s widget.'
}
