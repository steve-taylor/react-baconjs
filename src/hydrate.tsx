import React from 'react'
import ReactDOM from 'react-dom'

import PhaseContext from './context/PhaseContext'
import HydrationContextProvider from './context/HydrationContextProvider'
import keyFor from './keyFor'
import Phase from './Phase'
import {BaseProps, HydrationSingleWidgetData, Serializable, SerializableObject, WidgetType} from './types'

type HydrateOptions = {
    warnIfNotFound?: boolean
    warnIfAlreadyHydrated?: boolean
}

/**
 * Hydrate all dehydrated instances of the specified widget, returning a <code>Promise</code>.
 *
 * @param widget                          - a widget created with <code>widget</code>
 * @param [options]                       - options (see below)
 * @param [options.warnIfNotFound=false]  - whether or not to warn if the dehydrated widget is not found
 * @param [options.warnIfAlreadyHydrated] - whether or not to warn if the widget has already been hydrated
 */
export default function hydrate<PROPS extends SerializableObject, HYDRATION extends Serializable>(
    widget: WidgetType<PROPS>,
    {
        warnIfNotFound = false,
        warnIfAlreadyHydrated = true,
    }: HydrateOptions = {}
): void {
    /* istanbul ignore next */
    if (typeof window === 'undefined') {
        console.warn('Cannot hydrate a component outside a browser.')

        return
    }

    if (!widget.__widget_name__) {
        console.error('Cannot hydrate a component that isn‚Äôt a widget.')

        return
    }

    const {__WIDGET_DATA__: widgetData} = window

    if (!widgetData) {
        if (warnIfNotFound) {
            console.warn('No widgets to hydrate.')
        }

        return
    }

    const {__widget_name__: name} = widget
    const componentHydrations = widgetData[name] as HydrationSingleWidgetData<PROPS, HYDRATION>

    if (!componentHydrations) {
        if (warnIfNotFound) {
            console.warn(`No hydration data found for widget ‚Äú${name}‚Äù.`)
        }

        return
    }

    if (componentHydrations.hydrated) {
        if (warnIfAlreadyHydrated) {
            console.warn(`Widget ‚Äú${name}‚Äù is already hydrated.`)
        }

        return
    }

    for (const [instanceId, instanceHydration] of Object.entries(componentHydrations.instances)) {
        const element = document.getElementById(instanceId)

        if (!element) {
            console.error(`Cannot hydrate widget ‚Äú${name}‚Äù at mount point ‚Äú#${instanceId}‚Äù because the mount point wasn‚Äôt found.`)
            continue
        }

        try {
            const {props, hydration} = instanceHydration

            hydrateElement({widget, element, props, hydration})
        } catch (error) {
            console.error(`Widget ‚Äú${name}‚Äù at ‚Äú#${instanceId}‚Äù threw an error while hydrating.`)
            throw error
        }
    }

    componentHydrations.hydrated = true
}

// Hydrate a single element
function hydrateElement<PROPS extends BaseProps, HYDRATION extends Serializable>(
    {
        widget: Widget,
        element,
        props,
        hydration,
    }: {
        widget: WidgetType<PROPS>
        element: HTMLElement
        props: PROPS
        hydration: {[_widgetId: string]: HYDRATION}
    }
) {
    const {__widget_name__: name} = Widget

    // Debugging info
    /* istanbul ignore next */
    if (process.env.NODE_ENV === 'development') {
        console.info(`Hydrating component ‚Äú${name}‚Äù...`)
    }

    if (hydration) {
        let isHydrating = true

        // If we have initial data, hydrate the server-rendered component
        ReactDOM.hydrate(
            (
                <PhaseContext.Provider value={() => isHydrating ? Phase.hydration : null}>
                    <HydrationContextProvider<PROPS, HYDRATION>
                        value={(name: string, props: PROPS) => (
                            isHydrating
                                ? ({
                                    hydration: hydration[keyFor(name, props)],
                                    elementId: element.id,
                                })
                                : {
                                    hydration: undefined,
                                    elementId: undefined,
                                }
                        )}
                    >
                        <Widget {...(props as never)} />
                    </HydrationContextProvider>
                </PhaseContext.Provider>
            ),
            element,
            () => {
                isHydrating = false
            }
        )
    } else {
        // If we don't have initial data, render over the top of anything currently in the element.
        ReactDOM.render((
            <Widget {...(props as never)} />
        ), element)
    }

    // Debugging info
    /* istanbul ignore next */
    if (process.env.NODE_ENV === 'development') {
        console.info(`Widget ‚Äú${name}‚Äù hydrated üí¶ at`, element)
    }
}
