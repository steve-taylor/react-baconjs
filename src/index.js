// Common
export {default as widget} from './widget';
export {default as useWidgetState} from './use-widget-state';
export {default as Inject} from './inject';

// Server
export {default as renderToHtml} from './render-to-html';
export {default as StyledComponentsServerRenderer} from './styled-components-server-renderer';
export {SSR_TIMEOUT_ERROR} from './errors';

// Browser
export {default as hydrate} from './hydrate';
