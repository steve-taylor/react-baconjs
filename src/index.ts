// Common
export {default as widget} from './widget'
export {default as useWidgetState} from './useWidgetState'
export {Inject, Loading} from './inject'
export {default as connect} from './connect'
export {Serializable, SerializableObject, WidgetContext, WidgetData} from './types'

// Server
export {default as renderToHtml} from './renderToHtml'
export {default as StyledComponentsServerRenderer} from './StyledComponentsServerRenderer'
export {SSR_TIMEOUT_ERROR} from './errors'

// Browser
export {default as hydrate} from './hydrate'
