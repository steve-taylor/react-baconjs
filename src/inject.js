import propTypes from 'prop-types';

import useWidgetState from './use-widget-state';

export default function Inject({
    context,
    compare,
    children,
}) {
    const state = useWidgetState(context, compare);

    return state ? children(state) : null;
}

/* istanbul ignore next */
if (process.env.NODE_ENV === 'development') {
    Inject.propTypes = {

        /** React context for all instances of this component */
        context: propTypes.object.isRequired,

        /**
         * (Optional) A function that compares two consecutive widget states for equality,
         * or an array of widget property names to compare two consecutive
         * widget states
         */
        compare: propTypes.oneOfType([
            propTypes.func,
            propTypes.arrayOf(propTypes.string),
        ]),

        /** A render function that converts a widget state into React elements */
        children: propTypes.func.isRequired,
    };
}
