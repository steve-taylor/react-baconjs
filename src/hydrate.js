import React from 'react';
import ReactDOM from 'react-dom';

import {HydrationContext, PhaseContext, HYDRATION} from './context';
import keyFor from './key-for';

/**
 * Hydrate all dehydrated instances of the specified widget, returning a <code>Promise</code>.
 *
 * @param {Object}                   Widget                          - a widget created with <code>widget</code>
 * @param {Object}                   [options]                       - options (see below)
 * @param {boolean}                  [options.warnIfNotFound=false]  - whether or not to warn if the dehydrated widget is not found
 * @param {boolean}                  [options.warnIfAlreadyHydrated] - whether or not to warn if the widget has already been hydrated
 * @returns {undefined}
 */
export default function hydrate(
    Widget,
    {
        warnIfNotFound = false,
        warnIfAlreadyHydrated = true,
    } = {}
) {
    /* istanbul ignore next */
    if (typeof window === 'undefined') {
        console.warn('Cannot hydrate a component outside a browser.');

        return;
    }

    if (!Widget.__widget_name__) {
        console.error('Cannot hydrate a component that isn\'t a widget.');

        return;
    }

    const {__WIDGET_DATA__: widgetData} = window;

    if (!widgetData) {
        if (warnIfNotFound) {
            console.warn('No widgets to hydrate.');
        }

        return;
    }

    const {__widget_name__: name} = Widget;
    const componentHydrations = widgetData[name];

    if (!componentHydrations) {
        if (warnIfNotFound) {
            console.warn(`No hydration data found for widget "${name}".`);
        }

        return;
    }

    if (componentHydrations.hydrated) {
        if (warnIfAlreadyHydrated) {
            console.warn(`Widget "${name}" is already hydrated.`);
        }

        return;
    }

    for (const [elementId, elementHydration] of Object.entries(componentHydrations).filter(([elementId]) => elementId !== 'hydrated')) {
        // Shouldn't happen.
        if (elementHydration.hydrated) {
            console.error(`Widget "${name}" at mount point "#${elementId}" is already hydrated.`);
            continue;
        }

        const element = document.getElementById(elementId);

        if (!element) {
            console.error(`Cannot hydrate widget "${name}" at mount point "#${elementId}" because the mount point was not found.`);
            continue;
        }

        try {
            const {props, hydration} = elementHydration;

            hydrateElement({Widget, element, props, hydration});
            elementHydration.hydrated = true;
        } catch (error) {
            console.error(`Widget "${name}" at "#${elementId}" threw an error while hydrating.`);
            throw error;
        }
    }

    componentHydrations.hydrated = true;
}

// Hydrate a single element
function hydrateElement(
    {
        Widget,
        element,
        props,
        hydration,
    }
) {
    const {__widget_name__: name} = Widget;

    // Debugging info
    /* istanbul ignore next */
    if (process.env.NODE_ENV === 'development') {
        console.info(`Hydrating component "${name}"...`); // eslint-disable-line no-console
    }

    if (hydration) {
        let isHydrating = true;

        // If we have initial data, hydrate the server-rendered component
        ReactDOM.hydrate(
            (
                <PhaseContext.Provider value={() => isHydrating ? HYDRATION : null}>
                    <HydrationContext.Provider
                        value={(name, props) => (
                            isHydrating
                                ? ({
                                    hydration: hydration[keyFor(name, props)],
                                    elementId: element.id,
                                })
                                : {}
                        )}
                    >
                        <Widget {...props} />
                    </HydrationContext.Provider>
                </PhaseContext.Provider>
            ),
            element,
            () => {
                isHydrating = false;
            }
        );
    } else {
        // If we don't have initial data, render over the top of anything currently in the element.
        ReactDOM.render((
            <Widget {...props} />
        ), element);
    }

    // Debugging info
    /* istanbul ignore next */
    if (process.env.NODE_ENV === 'development') {
        console.info(`Widget "${name}" hydrated ðŸ’¦ at`, element); // eslint-disable-line no-console
    }
}
