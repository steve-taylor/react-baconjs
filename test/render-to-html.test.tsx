/// <reference path="global.d.ts" />

import {EventStream} from 'baconjs'
import noop from 'lodash/noop'
import React from 'react'
import {__DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS} from 'styled-components'
import {renderToHtml, StyledComponentsServerRenderer} from '../src'

import SimpleWidget1 from './data/hook/widgets/simple'
import NestedWidget1 from './data/hook/widgets/nested'
import NestedWithStylesWidget1 from './data/hook/widgets/nested-with-styles'

import SimpleWidget2 from './data/inject/widgets/simple'
import NestedWidget2 from './data/inject/widgets/nested'
import NestedWithStylesWidget2 from './data/inject/widgets/nested-with-styles'

import ThrowsDelayedErrorWidget from './data/inject/widgets/throws-delayed-error'
import ThrowsImmediateErrorWidget from './data/inject/widgets/throws-immediate-error'

import * as fetchBaseValue from './data/streams/fetch-base-value'
import * as fetchV from './data/streams/fetch-v'
import * as fetchW from './data/streams/fetch-w'

jest.mock('uuid', () => ({
    __esModule: true,
    v1: jest.fn(() => '0123456789abcdef'),
}))

const {StyleSheet} = __DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS // I ain't afraid of no ghosts

type FetchNumberSpy = jest.SpyInstance<EventStream<number>, []> | undefined

describe('renderToHtml(widget)', () => {
    [
        {
            name: 'useWidgetState()',
            SimpleWidget: SimpleWidget1,
            NestedWidget: NestedWidget1,
            NestedWithStylesWidget: NestedWithStylesWidget1,
        },
        {
            name: '<Inject />',
            SimpleWidget: SimpleWidget2,
            NestedWidget: NestedWidget2,
            NestedWithStylesWidget: NestedWithStylesWidget2,
        },
    ].forEach(({
        name,
        SimpleWidget,
        NestedWidget,
        NestedWithStylesWidget,
    }) => {
        describe(name, () => {
            describe('simple widget', () => {
                let fetchBaseValueSpy: FetchNumberSpy
                let body: string | undefined

                beforeEach(() => {
                    fetchBaseValueSpy = jest.spyOn(fetchBaseValue, 'default')
                })

                afterEach(() => {
                    body = undefined
                    fetchBaseValueSpy!.mockRestore()
                })

                describe('with mount point className', () => {
                    beforeEach(async () => {
                        body = await renderToHtml(<SimpleWidget power={4} />, {className: 'mount-point'})
                    })

                    test('renders correctly', () => {
                        expect({body}).toMatchSnapshot()
                    })

                    test('calls fetchBaseValue() once', () => {
                        expect(fetchBaseValueSpy!.mock.calls).toHaveLength(1)
                    })
                })

                describe('without mount point className', () => {
                    beforeEach(async () => {
                        body = await renderToHtml(<SimpleWidget power={4} />)
                    })

                    test('renders correctly', () => {
                        expect({body}).toMatchSnapshot()
                    })

                    test('calls fetchBaseValue() once', () => {
                        expect(fetchBaseValueSpy!.mock.calls).toHaveLength(1)
                    })
                })
            })

            describe('nested widget', () => {
                let fetchBaseValueSpy: FetchNumberSpy
                let fetchVSpy: FetchNumberSpy
                let fetchWSpy: FetchNumberSpy
                let body: string | undefined

                beforeEach(async () => {
                    fetchBaseValueSpy = jest.spyOn(fetchBaseValue, 'default')
                    fetchVSpy = jest.spyOn(fetchV, 'default')
                    fetchWSpy = jest.spyOn(fetchW, 'default')
                    body = await renderToHtml(<NestedWidget coefficient={9} />)
                })

                afterEach(() => {
                    body = undefined
                    fetchWSpy!.mockRestore()
                    fetchVSpy!.mockRestore()
                    fetchBaseValueSpy!.mockRestore()
                })

                test('renders correctly', () => {
                    expect({body}).toMatchSnapshot()
                })

                test('calls fetchBaseValue() once', () => {
                    expect(fetchBaseValueSpy!.mock.calls).toHaveLength(1)
                })

                test('calls fetchVSpy() once', () => {
                    expect(fetchVSpy!.mock.calls).toHaveLength(1)
                })

                test('calls fetchWSpy() once', () => {
                    expect(fetchWSpy!.mock.calls).toHaveLength(1)
                })
            })

            describe('styled nested widget', () => {
                beforeEach(() => {
                    // Pretend we're on a server
                    StyleSheet.reset(true)
                })

                afterEach(() => {
                    // Stop pretending
                    StyleSheet.reset()
                })

                test('renders correctly', async () => {
                    const renderer = new StyledComponentsServerRenderer()
                    const body = await renderToHtml(
                        <NestedWithStylesWidget coefficient={9} />,
                        {render: renderer.render}
                    )
                    const head = renderer.getStyleTags()

                    expect({head, body}).toMatchSnapshot()
                })
            })

            describe('regular component', () => {
                describe('with mount point className', () => {
                    test('renders as a regular element', async () => {
                        expect({body: await renderToHtml(<div>Hello ðŸ™‚</div>)}).toMatchSnapshot()
                    })
                })

                describe('without mount point className', () => {
                    test('renders as a regular element', async () => {
                        expect({body: await renderToHtml(<div>Hello ðŸ™‚</div>, {className: 'mount-point'})}).toMatchSnapshot()
                    })
                })
            })
        })
    })

    describe('stream(props$, hydration?) Observable produces an immediate error', () => {
        let originalConsoleError: typeof console.error | undefined
        let error: string | undefined

        beforeEach(async () => {
            originalConsoleError = console.error
            console.error = noop

            return renderToHtml(<ThrowsImmediateErrorWidget />).catch((e) => {
                error = e as string
            })
        })

        afterEach(() => {
            console.error = originalConsoleError!
        })

        test('propagates the error to the caller', () => {
            expect(error).toBe('Nope!')
        })
    })

    describe('stream(props$, hydration?) Observable produces a delayed error', () => {
        let originalConsoleError: typeof console.error | undefined
        let error: string | undefined

        beforeEach(async () => {
            originalConsoleError = console.error
            console.error = noop

            return renderToHtml(<ThrowsDelayedErrorWidget />).catch((e) => {
                error = e as string
            })
        })

        afterEach(() => {
            console.error = originalConsoleError!
        })

        test('propagates the error to the caller', () => {
            expect(error).toBe('Nope!')
        })
    })

    // Needs more tests
    describe('onMeta', () => {
        let accumulatedData: object = {}

        afterEach(() => {
            accumulatedData = {}
        })

        describe('child overrides parent; last sibling wins', () => {
            beforeEach(() => {
                return renderToHtml(
                    <NestedWidget1 coefficient={9} />,
                    {
                        onMeta(data) {
                            Object.assign(accumulatedData, data)
                        },
                    }
                )
            })

            test('accumulates the expected data', () => {
                expect(accumulatedData).toMatchSnapshot()
            })
        })

        describe('parent overrides child; first sibling wins', () => {
            beforeEach(() => {
                return renderToHtml(
                    <NestedWidget1 coefficient={9} />,
                    {
                        onMeta(meta: {maxAge: number}) {
                            accumulatedData = {
                                ...meta,
                                ...accumulatedData,
                            }
                        },
                    }
                )
            })

            test('accumulates the expected data', () => {
                expect(accumulatedData).toMatchSnapshot()
            })
        })
    })
})
