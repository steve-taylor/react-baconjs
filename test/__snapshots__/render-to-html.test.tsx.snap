// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderToHtml(widget) <Inject /> nested widget renders correctly 1`] = `
<head></head>

<body>
  <div id="0123456789abcdef">
    <section>
      <ul>
        <li>27</li>
        <li>72</li>
        <li>
          <section>625</section>
        </li>
      </ul>
    </section>
  </div>
  <script type="text/javascript">
    Object.assign(["__WIDGET_DATA__", "nested--injected", "instances", "0123456789abcdef"].reduce(function(a, b) {
      return a[b] = a[b] || {}
    }, window), {
      "props": {
        "coefficient": 9
      },
      "hydration": {
        "nested--injected--27152bfa9d152695b053e71d6ee08263979ed379d64fa16bfd8efdbf85ee0ffa": [3, 8],
        "simple--injected--f042b037451958bdc45147be8b9d937045a13c3523f59ba3514bebb4db43730c": 5
      }
    })
  </script>
</body>
`;

exports[`renderToHtml(widget) <Inject /> regular component with mount point className renders as a regular element 1`] = `
<head></head>

<body>
  <div>
    <div data-reactroot="">Hello ðŸ™‚</div>
  </div>
</body>
`;

exports[`renderToHtml(widget) <Inject /> regular component without mount point className renders as a regular element 1`] = `
<head></head>

<body>
  <div class="mount-point">
    <div data-reactroot="">Hello ðŸ™‚</div>
  </div>
</body>
`;

exports[`renderToHtml(widget) <Inject /> simple widget with mount point className renders correctly 1`] = `
<head></head>

<body>
  <div id="0123456789abcdef" class="mount-point">
    <section>625</section>
  </div>
  <script type="text/javascript">
    Object.assign(["__WIDGET_DATA__", "simple--injected", "instances", "0123456789abcdef"].reduce(function(a, b) {
      return a[b] = a[b] || {}
    }, window), {
      "props": {
        "power": 4
      },
      "hydration": {
        "simple--injected--f042b037451958bdc45147be8b9d937045a13c3523f59ba3514bebb4db43730c": 5
      }
    })
  </script>
</body>
`;

exports[`renderToHtml(widget) <Inject /> simple widget without mount point className renders correctly 1`] = `
<head></head>

<body>
  <div id="0123456789abcdef">
    <section>625</section>
  </div>
  <script type="text/javascript">
    Object.assign(["__WIDGET_DATA__", "simple--injected", "instances", "0123456789abcdef"].reduce(function(a, b) {
      return a[b] = a[b] || {}
    }, window), {
      "props": {
        "power": 4
      },
      "hydration": {
        "simple--injected--f042b037451958bdc45147be8b9d937045a13c3523f59ba3514bebb4db43730c": 5
      }
    })
  </script>
</body>
`;

exports[`renderToHtml(widget) <Inject /> styled nested widget renders correctly 1`] = `
<head>
  <style data-styled="ljOnAL gUgpWW" data-styled-version="4.4.1">
    /* sc-component-id: sc-htpNat */
    .ljOnAL {
      padding: 7px;
      background: #bbb;
    }

    /* sc-component-id: sc-bxivhb */
    .gUgpWW {
      margin: 7px;
      background: #666;
      color: #ddd;
    }
  </style>
</head>

<body>
  <div id="0123456789abcdef">
    <section class="sc-htpNat ljOnAL">
      <ul class="sc-bxivhb gUgpWW">
        <li>27</li>
        <li>72</li>
        <li>
          <section>625</section>
        </li>
      </ul>
    </section>
  </div>
  <script type="text/javascript">
    Object.assign(["__WIDGET_DATA__", "nested-with-styles--injected", "instances", "0123456789abcdef"].reduce(function(a, b) {
      return a[b] = a[b] || {}
    }, window), {
      "props": {
        "coefficient": 9
      },
      "hydration": {
        "nested-with-styles--injected--27152bfa9d152695b053e71d6ee08263979ed379d64fa16bfd8efdbf85ee0ffa": [3, 8],
        "simple--injected--f042b037451958bdc45147be8b9d937045a13c3523f59ba3514bebb4db43730c": 5
      }
    })
  </script>
</body>
`;

exports[`renderToHtml(widget) onMeta child overrides parent; last sibling wins accumulates the expected data 1`] = `
Object {
  "maxAge": 60,
}
`;

exports[`renderToHtml(widget) onMeta parent overrides child; first sibling wins accumulates the expected data 1`] = `
Object {
  "maxAge": 30,
}
`;

exports[`renderToHtml(widget) useWidgetState() nested widget renders correctly 1`] = `
<head></head>

<body>
  <div id="0123456789abcdef">
    <section>
      <ul>
        <li>27</li>
        <li>72</li>
        <li>
          <section>625</section>
        </li>
      </ul>
    </section>
  </div>
  <script type="text/javascript">
    Object.assign(["__WIDGET_DATA__", "nested--hooked", "instances", "0123456789abcdef"].reduce(function(a, b) {
      return a[b] = a[b] || {}
    }, window), {
      "props": {
        "coefficient": 9
      },
      "hydration": {
        "nested--hooked--27152bfa9d152695b053e71d6ee08263979ed379d64fa16bfd8efdbf85ee0ffa": [3, 8],
        "simple--hooked--f042b037451958bdc45147be8b9d937045a13c3523f59ba3514bebb4db43730c": 5
      }
    })
  </script>
</body>
`;

exports[`renderToHtml(widget) useWidgetState() regular component with mount point className renders as a regular element 1`] = `
<head></head>

<body>
  <div>
    <div data-reactroot="">Hello ðŸ™‚</div>
  </div>
</body>
`;

exports[`renderToHtml(widget) useWidgetState() regular component without mount point className renders as a regular element 1`] = `
<head></head>

<body>
  <div class="mount-point">
    <div data-reactroot="">Hello ðŸ™‚</div>
  </div>
</body>
`;

exports[`renderToHtml(widget) useWidgetState() simple widget with mount point className renders correctly 1`] = `
<head></head>

<body>
  <div id="0123456789abcdef" class="mount-point">
    <section>625</section>
  </div>
  <script type="text/javascript">
    Object.assign(["__WIDGET_DATA__", "simple--hooked", "instances", "0123456789abcdef"].reduce(function(a, b) {
      return a[b] = a[b] || {}
    }, window), {
      "props": {
        "power": 4
      },
      "hydration": {
        "simple--hooked--f042b037451958bdc45147be8b9d937045a13c3523f59ba3514bebb4db43730c": 5
      }
    })
  </script>
</body>
`;

exports[`renderToHtml(widget) useWidgetState() simple widget without mount point className renders correctly 1`] = `
<head></head>

<body>
  <div id="0123456789abcdef">
    <section>625</section>
  </div>
  <script type="text/javascript">
    Object.assign(["__WIDGET_DATA__", "simple--hooked", "instances", "0123456789abcdef"].reduce(function(a, b) {
      return a[b] = a[b] || {}
    }, window), {
      "props": {
        "power": 4
      },
      "hydration": {
        "simple--hooked--f042b037451958bdc45147be8b9d937045a13c3523f59ba3514bebb4db43730c": 5
      }
    })
  </script>
</body>
`;

exports[`renderToHtml(widget) useWidgetState() styled nested widget renders correctly 1`] = `
<head>
  <style data-styled="fDLULr gjMIGH" data-styled-version="4.4.1">
    /* sc-component-id: sc-bdVaJa */
    .fDLULr {
      padding: 7px;
      background: #bbb;
    }

    /* sc-component-id: sc-bwzfXH */
    .gjMIGH {
      margin: 7px;
      background: #666;
      color: #ddd;
    }
  </style>
</head>

<body>
  <div id="0123456789abcdef">
    <section class="sc-bdVaJa fDLULr">
      <ul class="sc-bwzfXH gjMIGH">
        <li>27</li>
        <li>72</li>
        <li>
          <section>625</section>
        </li>
      </ul>
    </section>
  </div>
  <script type="text/javascript">
    Object.assign(["__WIDGET_DATA__", "nested-with-styles--hooked", "instances", "0123456789abcdef"].reduce(function(a, b) {
      return a[b] = a[b] || {}
    }, window), {
      "props": {
        "coefficient": 9
      },
      "hydration": {
        "nested-with-styles--hooked--27152bfa9d152695b053e71d6ee08263979ed379d64fa16bfd8efdbf85ee0ffa": [3, 8],
        "simple--hooked--f042b037451958bdc45147be8b9d937045a13c3523f59ba3514bebb4db43730c": 5
      }
    })
  </script>
</body>
`;
